File: Assignment-7/stats.c
#include <stdio.h>
#include <stdlib.h>

// Comparison function for qsort to sort numbers in ascending order
int compare_numbers(const void *a, const void *b) {
    return (*(int*)a - *(int*)b);
}

// Calculates the mean (average) of the numbers
// Iterates through the array to sum all elements and divides by count
double compute_mean(int *numbers, int count) {
    double total_sum = 0.0;
    for (int i = 0; i < count; i++) {
        total_sum += numbers[i];
    }
    return total_sum / count;
}

// Calculates the median (middle value) of the sorted numbers
// Uses sorted array: for odd count, takes middle element; for even, averages two middle elements
double compute_median(int *numbers, int count) {
    // Assumes numbers is already sorted (via qsort in main)
    if (count % 2 == 1) {
        return numbers[count / 2];
    } else {
        return (numbers[count / 2 - 1] + numbers[count / 2]) / 2.0;
    }
}

// Calculates the mode (most frequent number(s))
// Algorithm: Iterates through sorted array, counts frequency of each number,
// keeps track of max frequency and collects all numbers with that frequency
void compute_mode(int *numbers, int count) {
    int max_frequency = 0, mode_frequency_count = 0;
    int *mode_values = (int*)malloc(count * sizeof(int));
    int i = 0;
    while (i < count) {
        int current_number = numbers[i];
        int frequency = 0;
        // Count frequency of current_number in sorted array
        while (i < count && numbers[i] == current_number) {
            frequency++;
            i++;
        }
        if (frequency > max_frequency) {
            // New max frequency found, reset mode list
            max_frequency = frequency;
            mode_frequency_count = 1;
            mode_values[0] = current_number;
        } else if (frequency == max_frequency) {
            // Add to list of modes if frequency matches max
            mode_values[mode_frequency_count] = current_number;
            mode_frequency_count++;
        }
    }
    printf("Mode: ");
    for (int j = 0; j < mode_frequency_count; j++) {
        printf("%d ", mode_values[j]);
    }
    printf("\n");
    free(mode_values);
}

int main() {
    int count;
    printf("Enter the number of elements: ");
    scanf("%d", &count);
    int *numbers = (int*)malloc(count * sizeof(int));
    printf("Enter the elements: ");
    for (int i = 0; i < count; i++) {
        scanf("%d", &numbers[i]);
    }
    // Sort array using qsort (stdlib's quicksort implementation)
    // Why: Sorting is required for median (middle element) and simplifies mode calculation
    qsort(numbers, count, sizeof(int), compare_numbers);
    printf("Mean: %.2f\n", compute_mean(numbers, count));
    printf("Median: %.2f\n", compute_median(numbers, count));
    compute_mode(numbers, count);
    free(numbers);
    return 0;
}

File: Assignment-7/stats.ml
(* Reads a list of integers from input *)
(* Uses recursion to collect 'count' number of inputs *)
let read_numbers count =
  let rec aux acc i =
    if i = count then List.rev acc
    else aux (read_int () :: acc) (i + 1)
  in
  aux [] 0

(* Calculates the mean (average) of the numbers *)
(* Algorithm: Uses fold_left to sum all elements, then divides by list length *)
let compute_mean numbers =
  let total_sum = List.fold_left (fun acc x -> acc + x) 0 numbers in
  float total_sum /. float (List.length numbers)

(* Calculates the median (middle value) of the sorted numbers *)
(* Algorithm: Sorts list, then for odd length takes middle element, for even averages two middle *)
let compute_median numbers =
  let sorted_numbers = List.sort compare numbers in
  let count = List.length sorted_numbers in
  if count mod 2 = 1 then float (List.nth sorted_numbers (count / 2))
  else
    let first_middle = List.nth sorted_numbers (count / 2 - 1) in
    let second_middle = List.nth sorted_numbers (count / 2) in
    (float first_middle +. float second_middle) /. 2.0

(* Calculates the mode (most frequent number(s)) *)
(* Algorithm: Sorts list, recursively counts frequencies in sorted list, 
   tracks max frequency and collects numbers with that frequency *)
let compute_mode numbers =
  let sorted_numbers = List.sort compare numbers in
  let rec find_modes max_frequency mode_values current_frequency current_number lst =
    match lst with
    | [] -> 
        if current_frequency > max_frequency then [current_number]
        else if current_frequency = max_frequency then current_number :: mode_values
        else mode_values
    | hd :: tl ->
        if hd = current_number then
          find_modes max_frequency mode_values (current_frequency + 1) current_number tl
        else
          let new_modes =
            if current_frequency > max_frequency then [current_number]
            else if current_frequency = max_frequency then current_number :: mode_values
            else mode_values
          in
          find_modes (max max_frequency current_frequency) new_modes 1 hd tl
  in
  match sorted_numbers with
  | [] -> []
  | hd :: tl -> find_modes 0 [] 1 hd tl

(* Main function to read input and display results *)
let () =
  print_string "Enter the number of elements: ";
  let count = read_int () in
  print_string "Enter the elements: ";
  let numbers = read_numbers count in
  Printf.printf "Mean: %.2f\n" (compute_mean numbers);
  Printf.printf "Median: %.2f\n" (compute_median numbers);
  let mode_values = compute_mode numbers in
  print_string "Mode: ";
  List.iter (fun x -> Printf.printf "%d " x) mode_values;
  print_newline ()